Add Custom Drink Entities Tami4 Edge

taltaub22 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
With tami4 you can create custom drinks (i.e specific temperature or specific volume) in the app.
This PR adds a button entity for each custom drink


## Type of change
- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
- Link to documentation pull request: WIP

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [X] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

Guy293 commented:
Hi, someone already started working on this: https://github.com/home-assistant/core/pull/123802

taltaub22 commented:
@Guy293 I didn't find it when I looked, probably because it's a draft.

I think I will close this one. Thanks!

taltaub22 committed:
Add drinks as button entities instead of using actions

taltaub22 commented:
@Guy293 
I took a look in #123802 and saw it wasn't updated in 3 weeks.
I've changed the design to add all the drinks as button entities.

I'm not sure my code is good as I just copied the existing button and added the `drink` attribute
If you have an idea on how to write a better code I'll glad to hear.

taltaub22 committed:
Merge branch 'dev' into tami4/add-service-prepare-drink

joostlek commented on homeassistant/components/tami4/button.py:
```diff
@@ -24,24 +26,40 @@ class Tami4EdgeButtonEntityDescription(ButtonEntityDescription):
+    device = await hass.async_add_executor_job(api.get_device)
+    drinks = device.drinks
+
+    drink_buttons = []
+    for drink in drinks:
+        drink_button = Tami4EdgeDrinkButtonEntityDescription(
+            key=drink.id,
+            name=drink.name,
+            press_fn=lambda api, drink: api.prepare_drink(drink),
+        )
+        drink_buttons.append(Tami4EdgeDrinkButton(api, drink_button, drink))
```
This could become a list comprehension

joostlek commented on homeassistant/components/tami4/button.py:
```diff
@@ -24,24 +26,40 @@ class Tami4EdgeButtonEntityDescription(ButtonEntityDescription):
-    async_add_entities(
-        Tami4EdgeButton(api, entity_description) for entity_description in BUTTONS
-    )
+    async_add_entities(drink_buttons)
+    async_add_entities([Tami4EdgeButton(api, BOIL_WATER_BUTTON)])
```
Please collect all entities you want to add in a list and add it all in once

taltaub22 committed:
Merge branch 'dev' into tami4/add-service-prepare-drink

taltaub22 committed:
Remove button extensions

joostlek commented on homeassistant/components/tami4/button.py:
```diff
@@ -24,24 +24,36 @@ class Tami4EdgeButtonEntityDescription(ButtonEntityDescription):
+            entity_description=Tami4EdgeButtonEntityDescription(
+                key=drink.id,
+                name=drink.name,
+                press_fn=lambda api, drink=drink: api.prepare_drink(drink),  # type: ignore[misc]
```
```suggestion
                press_fn=lambda api: api.prepare_drink(drink),
```

taltaub22 replied:
This change produces an error `B023 Function definition does not bind loop variable drink` because this can cause late binding

joostlek replied:
In that case you need to disregard my previous comments and we do need a second extension, sorry for that, didn't know this wasn't possible

taltaub22 replied:
I see that `type: ignored` is used a lot in the codebase. I'm thinking leaving it as is

taltaub22 commented on homeassistant/components/tami4/button.py:
```diff
@@ -24,24 +24,36 @@ class Tami4EdgeButtonEntityDescription(ButtonEntityDescription):
+            api=api,
+            entity_description=Tami4EdgeButtonEntityDescription(
+                key=drink.id,
+                name=drink.name,
```
This button prepare a drink to be poured out using the device.
The problem is that the drinks are created using the device app and the user can choose a name there

I can add the word "prepare" so for example for the drink called "cup of tea" the button name would be "prepare cup of tea"

joostlek replied:
That sounds awesome!

taltaub22 committed:
Add an extension to create new buttons

taltaub22 committed:
Merge remote-tracking branch 'origin/tami4/add-service-prepare-drink' into tami4/add-service-prepare-drink

# Conflicts:
#	homeassistant/components/tami4/button.py

taltaub22 committed:
Use translation key for buttons names

taltaub22 committed:
Change translation key wording

taltaub22 committed:
Merge branch 'dev' into tami4/add-service-prepare-drink

joostlek commented on homeassistant/components/tami4/button.py:
```diff
@@ -24,24 +26,46 @@ class Tami4EdgeButtonEntityDescription(ButtonEntityDescription):
-    async_add_entities(
-        Tami4EdgeButton(api, entity_description) for entity_description in BUTTONS
-    )
+    async_add_entities([Tami4EdgeButton(api, BOIL_WATER_BUTTON)])
+
+    device = await hass.async_add_executor_job(api.get_device)
+    drinks = device.drinks
+
+    drink_buttons = [
+        Tami4EdgeDrinkButton(
+            api=api,
+            entity_description=Tami4EdgeDrinkButtonEntityDescription(
+                key=drink.id,
+                translation_key="prepare_drink",
+                translation_placeholders={"drink_name": drink.name},
+                press_fn=lambda api, drink: api.prepare_drink(drink),
+            ),
+            drink=drink,
+        )
+        for drink in drinks
+    ]
+    async_add_entities(drink_buttons)
```
we still want to collect the entities in a list and call `async_add_entities` once

taltaub22 committed:
Call async_add_entities once

joostlek commented on homeassistant/components/tami4/button.py:
```diff
@@ -24,24 +26,48 @@ class Tami4EdgeButtonEntityDescription(ButtonEntityDescription):
+    device = await hass.async_add_executor_job(api.get_device)
+    drinks = device.drinks
+
+    drink_buttons = [
+        Tami4EdgeDrinkButton(
+            api=api,
+            entity_description=Tami4EdgeDrinkButtonEntityDescription(
+                key=drink.id,
+                translation_key="prepare_drink",
+                translation_placeholders={"drink_name": drink.name},
+                press_fn=lambda api, drink: api.prepare_drink(drink),
+            ),
+            drink=drink,
+        )
+        for drink in drinks
+    ]
+
+    buttons.extend(drink_buttons)
```
```suggestion
    buttons.extend(
        Tami4EdgeDrinkButton(
            api=api,
            entity_description=Tami4EdgeDrinkButtonEntityDescription(
                key=drink.id,
                translation_key="prepare_drink",
                translation_placeholders={"drink_name": drink.name},
                press_fn=lambda api, drink: api.prepare_drink(drink),
            ),
            drink=drink,
        )
        for drink in drinks
    )
```

taltaub22 committed:
Add icons

joostlek committed:
Update homeassistant/components/tami4/button.py

joostlek reviewed (approved):
Thanks
